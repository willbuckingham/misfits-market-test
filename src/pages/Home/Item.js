import React, { useState } from 'react'
import PropTypes from 'prop-types'
import {
  // IconButton,
  Typography,
  // Button,
  Card,
  CardMedia,
  CardContent,
  CardActions,
  Fab,
  TextField,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { 
  AddShoppingCart,
} from '@material-ui/icons'

const useStyles = makeStyles(({ spacing }) => ({
  card: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  cardContent: {
    flexGrow: 2,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  product: {
    lineHeight: 1.2,
  },
  priceContainer: {
    display: 'flex',
    alignItems: 'start',
    justifyContent: 'center',
    marginTop: spacing(1),
  },
  price: {
    fontWeight: 'bold',
  },
  msrp: {
    textDecoration: 'line-through',
  },
  cardActions: {
    display: 'flex',
    justifyContent: 'space-between',
    height: spacing(6),
    padding: spacing(2),
  },
}))

export const formatCurrency = v =>
  '$' + v.toFixed(2)

export const calcSavings = (msrp, price) => {
  const savingsExact = price < msrp ? (1 - price / msrp) * 100 : 0
  return Math.round(savingsExact / 5) * 5
}

const Item = props => {
  const classes = useStyles()

  const [ quantity, setQuantity ] = useState(1)

  const { id, product, price, msrp, soldOut } = props
  const savings = calcSavings(msrp, price)

  const onClickAdd = () => {
    props.onAdd({
      id,
      quantity,
    })
    setQuantity(1)
  }

  return (
    <Card className={classes.card}>

      <CardMedia
        component='img'
        image='/logo512.png'
        title={product}
        alt={product}
        height={200}
      />
      <CardContent className={classes.cardContent}>
        <Typography
          variant='h6'
          className={classes.product}
        >
          {product}
        </Typography>
        <div className={classes.priceContainer}>
          <Typography display='inline' variant='body1' color='primary' className={classes.price}>{formatCurrency(price)}</Typography>
          &nbsp;
          <Typography display='inline' variant='caption' color='textSecondary' className={classes.msrp} >{formatCurrency(msrp)}</Typography>
          &nbsp;
          { savings && <Typography display='inline' variant='caption' color='textSecondary'>{`${savings}% off MSRP`}</Typography> }
        </div>
      </CardContent>
      <CardActions className={classes.cardActions}>
        { soldOut ?
          <Typography color='error'>Sold Out!</Typography> :
          <>
            <TextField 
              value={quantity}
              onChange={ evt => setQuantity(evt.target.value) }
              label='Quantity'
              type='number'
              inputProps={{ min: 1, max: 99 }}
            />
            <Fab color='secondary' size='small' onClick={onClickAdd}>
              <AddShoppingCart/>
            </Fab>
          </>
        }
      </CardActions>
    </Card>
  )
}

Item.propTypes = {
  id: PropTypes.number.isRequired,
  product: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  msrp: PropTypes.number.isRequired,
  soldOut: PropTypes.bool.isRequired,
}

export default Item
