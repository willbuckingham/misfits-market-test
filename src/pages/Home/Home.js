import React from 'react'
import {
  Grid,
  Snackbar,
} from '@material-ui/core'
import Alert from '@material-ui/lab/Alert'

import logo from './logo.svg'
import './Home.css'

import Page from '../../components/Page'
import Item from './Item'

export class Home extends React.Component {
  
  // TENTATIVE STORE LOCATED AT PAGE LEVEL
  state = {
    market: {
      msg: '',
      data: {},
    },
    cart: [],
    notificationOpen: false,
  }

  componentDidMount() {
    fetch('https://applicant-dev.misfitsmarket.com/api/test/v1', {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    })
      .then(response => response.json())
      .then(data => {
        this.setState({
          market: data,
        })
      })
  }

  handleAdd = (entry) => {
    const existingEntry = this.state.cart.find(v => v.id === entry.id)

    this.setState({
      cart: existingEntry ?
        this.state.cart.map(v => v.id === entry.id ? { ...v, quantity: v.quantity + entry.quantity } : v) :
        this.state.cart.concat(entry),
      notificationOpen: true,
    })
  }

  handleNotifcationClose = () => {
    this.setState({
      notificationOpen: false,
    })
  }

  render () {
    const { market, cart, notificationOpen } = this.state
    
    const cartItems = market?.data?.items?.map(v => {
      const quantity = cart.find(c => v.id === c.id)?.quantity || 0

      return {
        ...v,
        quantity,
      }
    }).filter((v => v.quantity))

    return (
      <>
        <Page cartItems={cartItems}>
          { !market.msg ?
            <img src={logo} className='App-logo' alt='logo'/> :
            <Grid container spacing={3}>
              { market?.data?.items?.map(v => 
                <Grid item key={v.id} xl={2} lg={3} md={3} sm={4} xs={ 12 }>
                  <Item { ...v } onAdd={this.handleAdd} />
                </Grid>
              ) }
            </Grid>
          }
          <Snackbar open={notificationOpen} autoHideDuration={3000} onClose={this.handleNotifcationClose}>
            <Alert variant='filled' elevation={6} onClose={this.handleNotifcationClose} severity='success'>
              Item added to Cart!
            </Alert>
          </Snackbar>
        </Page>
      </>
    )
  }
}

export default Home
