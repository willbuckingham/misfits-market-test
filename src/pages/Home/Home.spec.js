import React from 'react'
import { shallow } from 'enzyme'

import Home from './Home'

describe('Home', () => {
  it('renders loading', () => {
    expect(shallow(<Home />)).toMatchSnapshot()
  })

  it('renders data', () => {
    const wrapper = shallow(<Home/>)

    wrapper.setState({
      market: {
        msg: 'success',
        data: {
          items: [{
            id: 1,
            product: 'Product',
            price: 2,
            msrp: 1,
            soldOut: false,
          }],
        },
      },
    })

    expect(wrapper).toMatchSnapshot()
  })

  xit('componentDidMount', () => {
    
  })
})
