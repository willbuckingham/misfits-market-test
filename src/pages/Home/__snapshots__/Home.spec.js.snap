// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders data 1`] = `
<Fragment>
  <Page
    cartItems={Array []}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={3}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        key="1"
        lg={3}
        md={3}
        sm={4}
        xl={2}
        xs={12}
      >
        <Item
          id={1}
          msrp={1}
          onAdd={[Function]}
          price={2}
          product="Product"
          soldOut={false}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Snackbar))
      autoHideDuration={3000}
      onClose={[Function]}
      open={false}
    >
      <WithStyles(ForwardRef(Alert))
        elevation={6}
        onClose={[Function]}
        severity="success"
        variant="filled"
      >
        Item added to Cart!
      </WithStyles(ForwardRef(Alert))>
    </WithStyles(ForwardRef(Snackbar))>
  </Page>
</Fragment>
`;

exports[`Home renders loading 1`] = `
<Fragment>
  <Page>
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <WithStyles(ForwardRef(Snackbar))
      autoHideDuration={3000}
      onClose={[Function]}
      open={false}
    >
      <WithStyles(ForwardRef(Alert))
        elevation={6}
        onClose={[Function]}
        severity="success"
        variant="filled"
      >
        Item added to Cart!
      </WithStyles(ForwardRef(Alert))>
    </WithStyles(ForwardRef(Snackbar))>
  </Page>
</Fragment>
`;
